include/envoy/router/router.h

```diff
@@ -226,6 +229,35 @@ class HashPolicy {
                                           const Http::HeaderMap& headers) const PURE;
 };
 
+class MetadataMatch {
+public:
+  virtual ~MetadataMatch() {}
+
+  /*
+   * @return const std::string& the name of the metadata key
+   */
+  virtual const std::string& name() const PURE;
+
+  /*
+   * @return const Envoy::HashedValue& the value for the metadata key
+   */
+  virtual const HashedValue& value() const PURE;
+};
+
+typedef std::shared_ptr<const MetadataMatch> MetadataMatchConstSharedPtr;
+
+class MetadataMatches {
+public:
+  virtual ~MetadataMatches() {}
+
+  /*
+   * @return std::vector<MetadataMatchConstSharedPtr>& a vector of
+   * metadata to be matched against upstream endpoints when load
+   * balancing, sorted lexically by name.
+   */
+  virtual const std::vector<MetadataMatchConstSharedPtr>& metadataMatches() const PURE;
```

Comment:
`metadataMatches()` or `metadataMatchers()`? The former seems to suggest we've already matched something.

Reply:
How about `metadataMatchCriteria()`? The object doesn't really do any matching itself.